fontir/src/variations.rs

```diff
@@ -113,6 +114,98 @@ impl VariationModel {
+    pub fn deltas<P, V>(
+        &self,
+        point_seqs: &HashMap<NormalizedLocation, Vec<P>>,
+    ) -> Result<HashMap<NormalizedLocation, Vec<V>>, DeltaError>
+    where
+        P: Copy + Default + Sub<P, Output = V>,
+        V: Copy + Mul<f64, Output = V> + Sub<V, Output = V>,
+    {
+        let Some(defaults) = point_seqs.get(&self.default) else {
+            return Err(DeltaError::DefaultUndefined);
+        };
+        if point_seqs.values().any(|pts| pts.len() != defaults.len()) {
+            return Err(DeltaError::InconsistentNumbersOfPoints);
+        }
+
+        let mut result: HashMap<NormalizedLocation, Vec<V>> = HashMap::new();
+        // self.locations is sorted such that[i] is only influenced by[i+1..N]
+        // so we know subsequent spins won't invalidate our delta if we go in the same order
+        for (loc, points, master_influences) in
+            self.locations
+                .iter()
+                .enumerate()
+                .filter_map(|(loc_idx, loc)| {
+                    point_seqs
+                        .get(loc)
+                        .map(|points| (loc, points, &self.delta_weights[loc_idx]))
+                })
+        {
+            let mut deltas = Vec::new();
+            for (idx, point) in points.iter().enumerate() {
+                let initial_vector: V = *point - Default::default();
+                deltas.push(
+                    // Find other masters that are active (have influence)
+                    // Any master with influence on us was processed already so we can get that masters
+                    // deltas from the results so far. If we subtract away all such influences what's
+                    // left is the delta to take us to point.
+                    master_influences
+                        .iter()
+                        .map(|(_, master_weight)| master_weight)
+                        .zip(&self.locations)
+                        .filter_map(|(master_weight, master_loc)| {
+                            let Some(master_deltas) = result.get(master_loc) else {
+                                return None;
+                            };
+                            let Some(delta) = master_deltas.get(idx) else {
+                                return None;
+                            };
+                            Some((delta, master_weight.into_inner()))
+                        })
+                        .fold(initial_vector, |acc, (other, other_weight)| {
+                            acc - if other_weight != 1.0 {
+                                *other * other_weight.into()
+                            } else {
+                                *other
+                            }
+                        }),
+                );
+            }
+            result.insert(loc.clone(), deltas);
+        }
+
+        Ok(result)
+    }
+}
+
+#[derive(Debug)]
+pub enum DeltaError {
```

Comment:
If we're going to use this as an actual public error type we should impl `Error`. I'm open to using some third-party crate like `thiserror` to make this easier?

Reply:
Oh good catch, will add.

Reply:
Done